### ISSUE TITLE
SCollection.transform should be public ?

### ISSUE THREAD

**Issue by jto** [spotify/scio#1473]
(Following a discussion with @kanterov which had more to it that just this question, but I'll let him submit another issue or just comment on this one).

There's a transform method on `SCollection` that I think may be interesting to make public.
It has the following signature:

```scala
private[scio] def transform[U](f: SCollection[T] => SCollection[U]): SCollection[U]
```

Making that public would give a clear pattern for structuring code. 
Similar to how people use `PTransform` in the Beam Java API, one could create a simple function of type `SCollection[T] => SCollection[U]` somewhere in their code and use it in their pipeline. 

```scala
// in file SharedTransforms.scala
def toJson: SCollection[String] => SCollection[Json] = _.map(...)

// In file MyAwesomePipeline.scala
sc.read(AvroIO[Foo](...))
  .map(_.jsonField)
  .transform(toJson)
  .write(BigQueryIO(...))
    //...
```

It would also give a clear pattern to share code between pipelines.

The refactoring on coder we made in 0.7 should give us the guarantee that the coders are properly set in `f`.

We could also overload the function with: 

```scala
def transform[U](name: String)(f: SCollection[T] => SCollection[U]): SCollection[U]
```
 Which would create an enclosing expandable named "box" in Dataflow's execution graph view.

@nevillelyh @regadas WDYT ?

**Comment by nevillelyh** [comment]
It wasn't since it's very hard to make sure `f: SCollection[T] => SCollection[U]` serializable. May we can create test cases for common ser/de issues and see if they can be worked around, e.g. `@transient`, `extends Serializable`, closure cleaning?

**Comment by jto** [comment]
How about we create a wrapper class that check the function is serializable ?

```scala
final case class STransform[T, U](f: SCollection[T] => SCollection[U]) {
  assert(isSerializable(f))
}
```

Then transform would be:

```scala
def transform[U](name: String)(f: STransform[T, U]): SCollection[U]
```

That way it should fail quickly in case of serialization isssue.

wdyt?

**Comment by regadas** [comment]
@jto I've been thinking about this and I think we can make the scope wider for this function and also provide the overloaded version with naming support. There's no need for the wrapper; `PTransform`'s are not serialized and `ClosureCleaner` is already being applied for the functions that are passed into `SCollection` functions.

With the widening of the scope i would also mark these functions with `@experimental` (we could potentially use `beam`'s) annotation do to potencial issues with `ClosureCleaner`.

**Comment by nevillelyh** [comment]
:+1: on trying it with examples. IIRC there're a couple of composite transforms on `SCollection` or `PairSCollectionFunctions` we can try those as well.

**Comment by regadas** [comment]
Closed via #1492
